File: HW/HW6/Solution.py
# Christopher Oh
# SJSU CS 146
# Homework 6

from typing import List


def threeSum(self, nums: List[int]) -> List[List[int]]:
    triplets = []
    if len(nums) < 3:
        print("Invalid array size")
        return triplets

    nums.sort()

    for a in range(0, len(nums) - 2):
        for b in range(a + 1, len(nums) - 1):
            third = -(nums[a] + nums[b])
            index = binary_search(nums, third, b + 1, len(nums) - 1)
            if index > 0:
                trio = [nums[a], nums[b], third]
                if trio not in triplets:
                    triplets.append(trio)

    return triplets


def binary_search(nums, target, left, right) -> int:
    while left <= right:
        mid = (left + right) / 2
        if nums[mid] > target:
            right = mid - 1
        elif nums[mid] < target:
            left = mid + 1
        else:
            return mid
    return -1


def main():
    print("Hi")


if __name__ == "__main__":
    main()

File: HW/HW6/Solution.java
import java.util.*;

public class Solution {
    public Solution() {}
    public List<List<Integer>> threeSum (int[] nums) {
        List<List<Integer>> answer = new ArrayList<>(); //list of all valid triplets
        if (nums.length < 3 || nums.length > 3000) {
            System.out.println("Invalid array size");
            return answer;
        }
        HashSet<ArrayList<Integer>> triplets = new HashSet<>();
        HashMap<Integer, Integer> vi = new HashMap<>(); //value index pairs of all elements in nums
        ArrayList<int[]> uniquePairs = new ArrayList<>(); //unique pairs


        //put all elements into vi
        //put all unique pairs into "pairs"
        for (int i = 0; i < nums.length; i++) {
            vi.put(nums[i], i);
            for (int j = i + 1; j < nums.length; j++) {
                int[] tempArr = {i, j};
                uniquePairs.add(tempArr);
            }
        }

        for (int[] duo : uniquePairs) {
            int third = -(nums[duo[0]] + nums[duo[1]]);
            if (vi.containsKey(third) && vi.get(third) != duo[0] && vi.get(third) != duo[1]) {
                ArrayList<Integer> temp = new ArrayList<>();
                temp.add(nums[duo[0]]);
                temp.add(nums[duo[1]]);
                temp.add(third);
                Collections.sort(temp);
                triplets.add(temp);
            }
        }

        
        for (ArrayList<Integer> set : triplets) {
            ArrayList<Integer> trioList = new ArrayList<>(set);
            triplets.add(trioList);
        }

        return answer;
    }
}
